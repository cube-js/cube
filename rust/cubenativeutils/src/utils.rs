use cubesql::CubeError;
use neon::prelude::*;
use std::fmt::Display;

#[inline(always)]
pub fn call_method<'a, AS>(
    cx: &mut impl Context<'a>,
    this: Handle<'a, JsFunction>,
    method_name: &str,
    args: AS,
) -> JsResult<'a, JsValue>
where
    AS: AsRef<[Handle<'a, JsValue>]>,
{
    let method: Handle<JsFunction> = this.get(cx, method_name)?;
    method.call(cx, this, args)
}

#[inline(always)]
pub fn bind_method<'a>(
    cx: &mut impl Context<'a>,
    fn_value: Handle<'a, JsFunction>,
    this: Handle<'a, JsValue>,
) -> JsResult<'a, JsValue> {
    call_method(cx, fn_value, "bind", [this])
}

// Extension trait to map abstract errors to CubeError
pub trait MapCubeErrExt<T> {
    fn map_cube_err(self, message: &str) -> Result<T, CubeError>;
}

impl<T, E: Display> MapCubeErrExt<T> for Result<T, E> {
    fn map_cube_err(self, message: &str) -> Result<T, CubeError> {
        self.map_err(|e| CubeError::user(format!("{}: {}", message, e)))
    }
}
