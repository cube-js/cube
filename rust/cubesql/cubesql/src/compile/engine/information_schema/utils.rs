use datafusion::arrow::array::{
    BooleanArray, BooleanBuilder, Int64Array, Int64Builder, StringArray, StringBuilder,
    UInt32Array, UInt32Builder,
};

pub fn new_string_array_with_placeholder(size: usize, default: Option<String>) -> StringArray {
    let mut builder = StringBuilder::new(size);

    if let Some(d) = default {
        for _ in 0..size {
            builder.append_value(d.as_str()).unwrap();
        }
    } else {
        for _ in 0..size {
            builder.append_null().unwrap();
        }
    };

    builder.finish()
}

pub fn new_int64_array_with_placeholder(size: usize, default: i64) -> Int64Array {
    let mut builder = Int64Builder::new(size);

    for _ in 0..size {
        builder.append_value(default).unwrap();
    }

    builder.finish()
}

pub fn new_uint32_array_with_placeholder(size: usize, default: u32) -> UInt32Array {
    let mut builder = UInt32Builder::new(size);

    for _ in 0..size {
        builder.append_value(default).unwrap();
    }

    builder.finish()
}

pub fn new_boolean_array_with_placeholder(size: usize, default: bool) -> BooleanArray {
    let mut builder = BooleanBuilder::new(size);

    for _ in 0..size {
        builder.append_value(default).unwrap();
    }

    builder.finish()
}
