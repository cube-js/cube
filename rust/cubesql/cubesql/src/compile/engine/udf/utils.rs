macro_rules! downcast_string_arg {
    ($ARG:expr, $NAME:expr, $T:ident) => {{
        $ARG.as_any()
            .downcast_ref::<datafusion::arrow::array::GenericStringArray<$T>>()
            .ok_or_else(|| {
                DataFusionError::Internal(format!(
                    "could not cast {} from {} to {}",
                    $NAME,
                    $ARG.data_type(),
                    type_name::<datafusion::arrow::array::GenericStringArray<$T>>()
                ))
            })?
    }};
}

macro_rules! downcast_boolean_arr {
    ($ARG:expr, $NAME:expr) => {{
        $ARG.as_any()
            .downcast_ref::<datafusion::arrow::array::BooleanArray>()
            .ok_or_else(|| {
                DataFusionError::Internal(format!(
                    "could not cast {} from {} to {}",
                    $NAME,
                    $ARG.data_type(),
                    type_name::<datafusion::arrow::array::BooleanArray>()
                ))
            })?
    }};
}

macro_rules! downcast_primitive_arg {
    ($ARG:expr, $NAME:expr, $T:ident) => {{
        $ARG.as_any()
            .downcast_ref::<datafusion::arrow::array::PrimitiveArray<$T>>()
            .ok_or_else(|| {
                DataFusionError::Internal(format!(
                    "could not cast {} from {} to {}",
                    $NAME,
                    $ARG.data_type(),
                    type_name::<$T>()
                ))
            })?
    }};
}

pub(crate) use downcast_boolean_arr;
pub(crate) use downcast_primitive_arg;
pub(crate) use downcast_string_arg;
