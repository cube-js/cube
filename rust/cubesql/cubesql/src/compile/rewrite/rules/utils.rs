use std::cmp::{max, min};

use chrono::{Datelike, NaiveDateTime, Timelike};
use datafusion::{physical_plan::aggregates::AggregateFunction, scalar::ScalarValue};

pub fn parse_granularity_string(granularity: &str, to_normalize: bool) -> Option<String> {
    if to_normalize {
        match granularity.to_lowercase().as_str() {
            "dow" | "doy" => Some("day".to_string()),
            "qtr" => Some("quarter".to_string()),
            _ => Some(granularity.to_lowercase()),
        }
    } else {
        match granularity.to_lowercase().as_str() {
            "qtr" => Some("quarter".to_string()),
            _ => Some(granularity.to_lowercase()),
        }
    }
}

pub fn parse_granularity(granularity: &ScalarValue, to_normalize: bool) -> Option<String> {
    match granularity {
        ScalarValue::Utf8(Some(granularity)) => {
            parse_granularity_string(&granularity, to_normalize)
        }
        _ => None,
    }
}

pub fn granularity_scalar_to_interval(granularity: &ScalarValue) -> Option<ScalarValue> {
    if let Some(granularity) = parse_granularity(granularity, false) {
        return granularity_to_interval(&granularity);
    }
    None
}

pub fn granularity_str_to_interval(granularity: &str) -> Option<ScalarValue> {
    if let Some(granularity) = parse_granularity_string(granularity, false) {
        return granularity_to_interval(&granularity);
    }
    None
}

fn granularity_to_interval(granularity: &str) -> Option<ScalarValue> {
    let interval = match granularity.to_lowercase().as_str() {
        "second" => ScalarValue::IntervalDayTime(Some(1000)),
        "minute" => ScalarValue::IntervalDayTime(Some(60000)),
        "hour" => ScalarValue::IntervalDayTime(Some(3600000)),
        "day" => ScalarValue::IntervalDayTime(Some(4294967296)),
        "week" => ScalarValue::IntervalDayTime(Some(30064771072)),
        "month" => ScalarValue::IntervalYearMonth(Some(1)),
        "quarter" => ScalarValue::IntervalYearMonth(Some(3)),
        "year" => ScalarValue::IntervalYearMonth(Some(12)),
        _ => return None,
    };
    Some(interval)
}

pub fn min_max_granularity(
    left: &ScalarValue,
    right: &ScalarValue,
    is_max: bool,
    week_as_day: Option<bool>,
) -> Option<String> {
    let left = parse_granularity(left, false)?;
    let right = parse_granularity(right, false)?;

    let left = granularity_str_to_int_order(&left, week_as_day)?;
    let right = granularity_str_to_int_order(&right, week_as_day)?;

    let result = if is_max {
        max(left, right)
    } else {
        min(left, right)
    };
    granularity_int_order_to_str(result, week_as_day)
}

pub fn granularity_str_to_int_order(granularity: &str, week_as_day: Option<bool>) -> Option<i32> {
    match granularity.to_lowercase().as_str() {
        "second" => Some(0),
        "minute" => Some(1),
        "hour" => Some(2),
        "day" => Some(3),
        // Week-month offsets may lead to incorrect results. `week_as_day` controls
        // the result of week granularity conversion.
        "week" => match week_as_day {
            Some(true) => Some(3),
            Some(false) => Some(4),
            None => None,
        },
        "month" => Some(5),
        "quarter" => Some(6),
        "year" => Some(7),
        _ => None,
    }
}

fn granularity_int_order_to_str(granularity: i32, week_as_day: Option<bool>) -> Option<String> {
    match granularity {
        0 => Some("second"),
        1 => Some("minute"),
        2 => Some("hour"),
        3 => Some("day"),
        // Week-month offsets may lead to incorrect results. `week_as_day` controls
        // the result of week granularity conversion.
        4 => match week_as_day {
            Some(true) => Some("day"),
            Some(false) => Some("week"),
            None => None,
        },
        5 => Some("month"),
        6 => Some("quarter"),
        7 => Some("year"),
        _ => None,
    }
    .map(|g| g.to_string())
}

pub fn negated_cube_filter_op(op: &str) -> Option<&'static str> {
    macro_rules! define_op_eq {
        ($($EXPR:expr => $NEG:expr,)*) => {
            match op {
                $(
                    $EXPR => $NEG,
                    $NEG => $EXPR,
                )*
                _ => return None,
            }
        }
    }

    let negated = define_op_eq![
        "equals" => "notEquals",
        "contains" => "notContains",
        "startsWith" => "notStartsWith",
        "endsWith" => "notEndsWith",
        "gt" => "lte",
        "lt" => "gte",
        "set" => "notSet",
        "inDateRange" => "notInDateRange",
    ];

    Some(negated)
}

pub fn reaggragate_fun(cube_fun: &str) -> Option<AggregateFunction> {
    Some(match cube_fun {
        "count" | "sum" => AggregateFunction::Sum,
        "min" => AggregateFunction::Min,
        "max" => AggregateFunction::Max,
        _ => return None,
    })
}

pub fn is_literal_date_trunced(ns: i64, granularity: &str) -> Option<bool> {
    let granularity = parse_granularity_string(granularity, false)?;
    let ns_in_seconds = 1_000_000_000;
    if ns % ns_in_seconds > 0 {
        return Some(false);
    }
    let seconds = ns / ns_in_seconds;
    let dt = NaiveDateTime::from_timestamp_opt(seconds, 0)?;

    let is_minute_trunced = |dt: NaiveDateTime| dt.second() == 0;
    let is_hour_trunced = |dt| is_minute_trunced(dt) && dt.minute() == 0;
    let is_day_trunced = |dt| is_hour_trunced(dt) && dt.hour() == 0;
    let is_week_trunced = |dt| is_day_trunced(dt) && dt.weekday().num_days_from_monday() == 0;
    let is_month_trunced = |dt| is_day_trunced(dt) && dt.day() == 1;
    let is_quarter_trunced = |dt| is_month_trunced(dt) && dt.month0() % 3 == 0;
    let is_year_trunced = |dt| is_month_trunced(dt) && dt.month() == 1;

    Some(match granularity.as_str() {
        "second" => true,
        "minute" => is_minute_trunced(dt),
        "hour" => is_hour_trunced(dt),
        "day" => is_day_trunced(dt),
        "week" => is_week_trunced(dt),
        "month" => is_month_trunced(dt),
        "quarter" => is_quarter_trunced(dt),
        "year" => is_year_trunced(dt),
        _ => return None,
    })
}
