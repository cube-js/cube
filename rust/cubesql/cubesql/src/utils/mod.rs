use crate::compile::rewrite::{analysis::LogicalPlanAnalysis, LogicalPlanLanguage};
use datafusion::scalar::ScalarValue;
use egg::EGraph;
use sha2::{Digest, Sha256};
use std::{
    collections::HashMap,
    convert::TryInto,
    hash::{Hash, Hasher},
};

pub struct ShaHasher {
    hasher: Sha256,
}

impl Hasher for ShaHasher {
    #[inline]
    fn finish(&self) -> u64 {
        let mut result = [0; 32];
        result.copy_from_slice(&self.hasher.clone().finalize());
        u64::from_le_bytes(result[0..8].try_into().unwrap())
    }

    #[inline]
    fn write(&mut self, bytes: &[u8]) {
        self.hasher.update(bytes);
    }
}

impl ShaHasher {
    pub fn new() -> Self {
        Self {
            hasher: Sha256::new(),
        }
    }

    pub fn take_hasher(self) -> Sha256 {
        self.hasher
    }
}

pub fn egraph_hash(
    egraph: &EGraph<LogicalPlanLanguage, LogicalPlanAnalysis>,
    params: Option<&HashMap<usize, ScalarValue>>,
) -> [u8; 32] {
    let mut hasher = ShaHasher::new();
    for class in egraph.classes() {
        class.id.hash(&mut hasher);
        class.nodes.len().hash(&mut hasher);
        for node in &class.nodes {
            node.hash(&mut hasher);
        }
    }
    if let Some(params) = params {
        for (k, v) in params.iter() {
            k.hash(&mut hasher);
            v.hash(&mut hasher);
        }
    }
    hasher.take_hasher().finalize().into()
}
