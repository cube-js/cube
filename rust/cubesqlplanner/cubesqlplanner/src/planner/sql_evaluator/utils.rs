use super::{MemberEvaluator, MemberEvaluatorFactory};
use crate::cube_bridge::evaluator::CubeEvaluator;
use crate::cube_bridge::measure_definition::MeasureDefinition;
use crate::cube_bridge::memeber_sql::MemberSql;
use crate::planner::query_tools::QueryTools;
use cubenativeutils::CubeError;
use std::any::Any;
use std::rc::Rc;

pub fn evaluate_sql(
    tools: Rc<QueryTools>,
    sql: Rc<dyn MemberSql>,
    deps: &Vec<Rc<dyn MemberEvaluator>>,
) -> Result<String, CubeError> {
    let args = deps
        .iter()
        .map(|dep| dep.eveluate(tools.clone()))
        .collect::<Result<Vec<_>, _>>()?;
    sql.call(args)
}
