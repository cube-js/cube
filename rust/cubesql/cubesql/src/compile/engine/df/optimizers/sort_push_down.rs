use std::{collections::HashMap, sync::Arc};

use datafusion::{
    error::{DataFusionError, Result},
    logical_plan::{
        plan::{
            Aggregate, CrossJoin, Distinct, Join, Limit, Projection, Sort, Subquery, Union, Window,
        },
        Column, DFSchema, Expr, Filter, LogicalPlan,
    },
    optimizer::optimizer::{OptimizerConfig, OptimizerRule},
};

use super::utils::{get_schema_columns, is_column_expr, plan_has_projections, rewrite};

/// Sort Push Down optimizer rule pushes ORDER BY clauses consisting of specific,
/// mostly simple, expressions down the plan, all the way to the Projection
/// closest to TableScan. This is beneficial for CubeScans when some of the Projections
/// on the way contain post-processing operations and cannot be pushed down.
#[derive(Default)]
pub struct SortPushDown {}

impl SortPushDown {
    #[allow(missing_docs)]
    pub fn new() -> Self {
        Self {}
    }
}

impl OptimizerRule for SortPushDown {
    fn optimize(
        &self,
        plan: &LogicalPlan,
        optimizer_config: &OptimizerConfig,
    ) -> Result<LogicalPlan> {
        sort_push_down(self, plan, None, optimizer_config)
    }

    fn name(&self) -> &str {
        "__cube__sort_push_down"
    }
}

/// Recursively optimizes plan, collecting sort expressions that can possibly be pushed down.
/// Only the topmost sort expression is kept when one pushes through another.
fn sort_push_down(
    optimizer: &SortPushDown,
    plan: &LogicalPlan,
    sort_expr: Option<Vec<Expr>>,
    optimizer_config: &OptimizerConfig,
) -> Result<LogicalPlan> {
    match plan {
        LogicalPlan::Projection(Projection {
            expr,
            input,
            schema,
            alias,
        }) => {
            // Sort can be pushed down to projection, however we only map specific expressions.
            // Complex expressions can't be pushed down, so if there are any, Sort is issued
            // before the projection.
            if plan_has_projections(input) {
                if let Some(sort_expr) = &sort_expr {
                    let rewrite_map = rewrite_map_for_projection(expr, schema);
                    if let Some(new_sort_expr) = sort_expr
                        .iter()
                        .map(|expr| match expr {
                            Expr::Sort {
                                expr,
                                asc,
                                nulls_first,
                            } => Ok(if is_column_expr(expr) {
                                rewrite(expr, &rewrite_map)?.map(|expr| Expr::Sort {
                                    expr: Box::new(expr),
                                    asc: asc.clone(),
                                    nulls_first: nulls_first.clone(),
                                })
                            } else {
                                None
                            }),
                            _ => Err(DataFusionError::Internal(
                                "Unable to optimize plan: sort contains non-sort expressions"
                                    .to_string(),
                            )),
                        })
                        .collect::<Result<Option<_>>>()?
                    {
                        return Ok(LogicalPlan::Projection(Projection {
                            expr: expr.clone(),
                            input: Arc::new(sort_push_down(
                                optimizer,
                                input,
                                Some(new_sort_expr),
                                optimizer_config,
                            )?),
                            schema: schema.clone(),
                            alias: alias.clone(),
                        }));
                    }
                }
            }

            issue_sort(
                sort_expr,
                LogicalPlan::Projection(Projection {
                    expr: expr.clone(),
                    input: Arc::new(sort_push_down(optimizer, input, None, optimizer_config)?),
                    schema: schema.clone(),
                    alias: alias.clone(),
                }),
            )
        }
        LogicalPlan::Filter(Filter { predicate, input }) => {
            // Sort can be pushed down Filter, and while it may seem weird to do that
            // after doing the exact opposite in `FilterPushDown`, this may allow the sort
            // to push through some complex filters, ultimately reaching CubeScan.
            Ok(LogicalPlan::Filter(Filter {
                predicate: predicate.clone(),
                input: Arc::new(sort_push_down(
                    optimizer,
                    input,
                    sort_expr,
                    optimizer_config,
                )?),
            }))
        }
        LogicalPlan::Window(Window {
            input,
            window_expr,
            schema,
        }) => {
            // Sort can't be pushed down Window, but we can optimize its input.
            issue_sort(
                sort_expr,
                LogicalPlan::Window(Window {
                    input: Arc::new(sort_push_down(optimizer, input, None, optimizer_config)?),
                    window_expr: window_expr.clone(),
                    schema: schema.clone(),
                }),
            )
        }
        LogicalPlan::Aggregate(Aggregate {
            input,
            group_expr,
            aggr_expr,
            schema,
        }) => {
            // It may be unsafe to push Sort down Aggregate; optimize just the input.
            issue_sort(
                sort_expr,
                LogicalPlan::Aggregate(Aggregate {
                    input: Arc::new(sort_push_down(optimizer, input, None, optimizer_config)?),
                    group_expr: group_expr.clone(),
                    aggr_expr: aggr_expr.clone(),
                    schema: schema.clone(),
                }),
            )
        }
        LogicalPlan::Sort(Sort { expr, input }) => {
            // When encountering Sort, drop it from the plan, keeping the expr.
            // If we already have an expr, however, then there was a sort above which
            // would override this sort expression; drop the new one in such case.
            sort_push_down(
                optimizer,
                input,
                Some(sort_expr.unwrap_or(expr.clone())),
                optimizer_config,
            )
        }
        LogicalPlan::Join(Join {
            left,
            right,
            on,
            join_type,
            join_constraint,
            schema,
            null_equals_null,
        }) => {
            // DataFusion preserves the sorting of the joined plans, prioritizing left side.
            // Taking this into account, we can push Sort down the left plan if Sort references
            // columns just from the left side.
            // TODO: check if this is still the case with multiple target partitions
            if let Some(some_sort_expr) = &sort_expr {
                let left_columns = get_schema_columns(left.schema());
                if some_sort_expr.iter().all(|expr| {
                    if let Expr::Sort { expr, .. } = expr {
                        if let Expr::Column(column) = expr.as_ref() {
                            return left_columns.contains(column);
                        }
                    }
                    false
                }) {
                    return Ok(LogicalPlan::Join(Join {
                        left: Arc::new(sort_push_down(
                            optimizer,
                            left,
                            sort_expr,
                            optimizer_config,
                        )?),
                        right: Arc::new(sort_push_down(optimizer, right, None, optimizer_config)?),
                        on: on.clone(),
                        join_type: join_type.clone(),
                        join_constraint: join_constraint.clone(),
                        schema: schema.clone(),
                        null_equals_null: null_equals_null.clone(),
                    }));
                }
            }

            issue_sort(
                sort_expr,
                LogicalPlan::Join(Join {
                    left: Arc::new(sort_push_down(optimizer, left, None, optimizer_config)?),
                    right: Arc::new(sort_push_down(optimizer, right, None, optimizer_config)?),
                    on: on.clone(),
                    join_type: join_type.clone(),
                    join_constraint: join_constraint.clone(),
                    schema: schema.clone(),
                    null_equals_null: null_equals_null.clone(),
                }),
            )
        }
        LogicalPlan::CrossJoin(CrossJoin {
            left,
            right,
            schema,
        }) => {
            // See `LogicalPlan::Join` notes above.
            if let Some(some_sort_expr) = &sort_expr {
                let left_columns = get_schema_columns(left.schema());
                if some_sort_expr.iter().all(|expr| {
                    if let Expr::Sort { expr, .. } = expr {
                        if let Expr::Column(column) = expr.as_ref() {
                            return left_columns.contains(column);
                        }
                    }
                    false
                }) {
                    return Ok(LogicalPlan::CrossJoin(CrossJoin {
                        left: Arc::new(sort_push_down(
                            optimizer,
                            left,
                            sort_expr,
                            optimizer_config,
                        )?),
                        right: Arc::new(sort_push_down(optimizer, right, None, optimizer_config)?),
                        schema: schema.clone(),
                    }));
                }
            }

            issue_sort(
                sort_expr,
                LogicalPlan::CrossJoin(CrossJoin {
                    left: Arc::new(sort_push_down(optimizer, left, None, optimizer_config)?),
                    right: Arc::new(sort_push_down(optimizer, right, None, optimizer_config)?),
                    schema: schema.clone(),
                }),
            )
        }
        LogicalPlan::Union(Union {
            inputs,
            schema,
            alias,
        }) => {
            // Union randomizes sorting, so Sort can't be pushed down.
            issue_sort(
                sort_expr,
                LogicalPlan::Union(Union {
                    inputs: inputs
                        .iter()
                        .map(|input| sort_push_down(optimizer, input, None, optimizer_config))
                        .collect::<Result<_>>()?,
                    schema: schema.clone(),
                    alias: alias.clone(),
                }),
            )
        }
        plan @ LogicalPlan::TableScan(_) | plan @ LogicalPlan::EmptyRelation(_) => {
            // TableScan or EmptyRelation's as far as we can push our sort expression.
            issue_sort(sort_expr, plan.clone())
        }
        LogicalPlan::Limit(Limit { skip, fetch, input }) => {
            // Pushing down Sort to Limit will affect the results; issue the sort expression.
            issue_sort(
                sort_expr,
                LogicalPlan::Limit(Limit {
                    skip: skip.clone(),
                    fetch: fetch.clone(),
                    input: Arc::new(sort_push_down(optimizer, input, None, optimizer_config)?),
                }),
            )
        }
        LogicalPlan::Subquery(Subquery {
            subqueries,
            input,
            schema,
        }) => {
            // TODO: Pushing Sort down Subquery?
            issue_sort(
                sort_expr,
                LogicalPlan::Subquery(Subquery {
                    subqueries: subqueries
                        .iter()
                        .map(|subquery| sort_push_down(optimizer, subquery, None, optimizer_config))
                        .collect::<Result<_>>()?,
                    input: Arc::new(sort_push_down(optimizer, input, None, optimizer_config)?),
                    schema: schema.clone(),
                }),
            )
        }
        LogicalPlan::Distinct(Distinct { input }) => {
            // Distinct randomizes the sorting; issue the sort expression.
            issue_sort(
                sort_expr,
                LogicalPlan::Distinct(Distinct {
                    input: Arc::new(sort_push_down(optimizer, input, None, optimizer_config)?),
                }),
            )
        }
        other => {
            // The rest of the plans have no inputs to optimize, can't have sort expressions
            // be pushed down them, or it makes no sense to optimize them.
            issue_sort(sort_expr, other.clone())
        }
    }
}

/// Generates a rewrite map for projection, taking qualified and unqualified fields into account.
/// Only simple realiasing expressions are mapped, with specific exceptions; more complex
/// projection expressions might produce complex sort expressions which cannot be pushed down to CubeScan,
/// and will block other nodes: those are mapped as `None` to explicitly mark them as non-mappable.
/// Extend this on case-by-case basis.
fn rewrite_map_for_projection(
    exprs: &Vec<Expr>,
    schema: &Arc<DFSchema>,
) -> HashMap<Column, Option<Expr>> {
    schema
        .fields()
        .iter()
        .zip(exprs)
        .flat_map(|(field, expr)| {
            // Aliases are never part of ORDER BY clause so they must be removed
            let expr = match expr {
                Expr::Alias(expr, _) => expr,
                expr @ _ => expr,
            };

            let expr = match expr {
                // We always expand simple realiasing expressions
                expr @ Expr::Column(_) => Some(expr.clone()),
                _ => None,
            };

            // Duplicate fields for projections without an alias
            // will be dropped while collecting as HashMap
            vec![
                (field.qualified_column(), expr.clone()),
                (field.unqualified_column(), expr),
            ]
        })
        .collect()
}

/// Issues a Sort containing the provided input if the provided `sort_expr` is `Some`;
/// otherwise, issues the provided input instead.
fn issue_sort(sort_expr: Option<Vec<Expr>>, input: LogicalPlan) -> Result<LogicalPlan> {
    if let Some(sort_expr) = sort_expr {
        return Ok(LogicalPlan::Sort(Sort {
            expr: sort_expr,
            input: Arc::new(input),
        }));
    }
    Ok(input)
}

#[cfg(test)]
mod tests {
    use super::{
        super::utils::{make_sample_table, sample_table},
        *,
    };
    use datafusion::logical_plan::{col, JoinType, LogicalPlanBuilder};

    fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
        let rule = SortPushDown::new();
        rule.optimize(plan, &OptimizerConfig::new())
    }

    fn assert_optimized_plan_eq(plan: LogicalPlan, expected: &str) {
        let optimized_plan = optimize(&plan).expect("failed to optimize plan");
        let formatted_plan = format!("{:?}", optimized_plan);
        assert_eq!(formatted_plan, expected);
    }

    fn sort(expr: Expr, asc: bool, nulls_first: bool) -> Expr {
        Expr::Sort {
            expr: Box::new(expr),
            asc,
            nulls_first,
        }
    }

    #[test]
    fn test_sort_down_projection() -> Result<()> {
        let plan = LogicalPlanBuilder::from(sample_table()?)
            .project(vec![col("c1"), col("c2"), col("c3")])?
            .project_with_alias(
                vec![col("c1").alias("n1"), col("c2"), col("c3").alias("n2")],
                Some("t2".to_string()),
            )?
            .sort(vec![
                sort(col("t2.c2"), true, false),
                sort(col("t2.n2"), false, true),
            ])?
            .build()?;

        let expected = "\
              Projection: #t1.c1 AS n1, #t1.c2, #t1.c3 AS n2, alias=t2\
            \n  Sort: #t1.c2 ASC NULLS LAST, #t1.c3 DESC NULLS FIRST\
            \n    Projection: #t1.c1, #t1.c2, #t1.c3\
            \n      TableScan: t1 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);
        Ok(())
    }

    #[test]
    fn test_sort_down_multiple_projections() -> Result<()> {
        let plan = LogicalPlanBuilder::from(sample_table()?)
            .project(vec![col("c1"), col("c2"), col("c3")])?
            .project_with_alias(
                vec![col("c1").alias("n1"), col("c2"), col("c3").alias("n2")],
                Some("t2".to_string()),
            )?
            .project_with_alias(
                vec![col("n1").alias("n3"), col("c2").alias("n4"), col("n2")],
                Some("t3".to_string()),
            )?
            .project_with_alias(
                vec![col("n3"), col("n4"), col("n2")],
                Some("t4".to_string()),
            )?
            .sort(vec![
                sort(col("t4.n4"), true, false),
                sort(col("t4.n2"), false, true),
            ])?
            .build()?;

        let expected = "\
              Projection: #t3.n3, #t3.n4, #t3.n2, alias=t4\
            \n  Projection: #t2.n1 AS n3, #t2.c2 AS n4, #t2.n2, alias=t3\
            \n    Projection: #t1.c1 AS n1, #t1.c2, #t1.c3 AS n2, alias=t2\
            \n      Sort: #t1.c2 ASC NULLS LAST, #t1.c3 DESC NULLS FIRST\
            \n        Projection: #t1.c1, #t1.c2, #t1.c3\
            \n          TableScan: t1 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);
        Ok(())
    }

    #[test]
    fn test_sort_down_sort() -> Result<()> {
        let plan = LogicalPlanBuilder::from(sample_table()?)
            .project(vec![col("c1"), col("c2"), col("c3")])?
            .project_with_alias(
                vec![col("c1").alias("n1"), col("c2"), col("c3").alias("n2")],
                Some("t2".to_string()),
            )?
            .sort(vec![sort(col("t2.n1"), false, false)])?
            .project_with_alias(
                vec![col("n1").alias("n3"), col("c2").alias("n4"), col("n2")],
                Some("t3".to_string()),
            )?
            .sort(vec![sort(col("t3.n2"), true, true)])?
            .project_with_alias(
                vec![col("n3"), col("n4"), col("n2")],
                Some("t4".to_string()),
            )?
            .sort(vec![
                sort(col("t4.n4"), true, false),
                sort(col("t4.n2"), false, true),
            ])?
            .build()?;

        let expected = "\
              Projection: #t3.n3, #t3.n4, #t3.n2, alias=t4\
            \n  Projection: #t2.n1 AS n3, #t2.c2 AS n4, #t2.n2, alias=t3\
            \n    Projection: #t1.c1 AS n1, #t1.c2, #t1.c3 AS n2, alias=t2\
            \n      Sort: #t1.c2 ASC NULLS LAST, #t1.c3 DESC NULLS FIRST\
            \n        Projection: #t1.c1, #t1.c2, #t1.c3\
            \n          TableScan: t1 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);
        Ok(())
    }

    #[test]
    fn test_sort_down_join() -> Result<()> {
        let plan = LogicalPlanBuilder::from(
            LogicalPlanBuilder::from(make_sample_table("j1", vec!["key", "c1"])?)
                .project(vec![col("key"), col("c1")])?
                .build()?,
        )
        .join(
            &LogicalPlanBuilder::from(make_sample_table("j2", vec!["key", "c2"])?)
                .project(vec![col("key"), col("c2")])?
                .build()?,
            JoinType::Inner,
            (
                vec![Column::from_name("key")],
                vec![Column::from_name("key")],
            ),
        )?
        .project(vec![col("j1.c1"), col("j2.c2")])?
        .sort(vec![sort(col("j1.c1"), true, false)])?
        .build()?;

        let expected = "\
              Projection: #j1.c1, #j2.c2\
            \n  Inner Join: #j1.key = #j2.key\
            \n    Sort: #j1.c1 ASC NULLS LAST\
            \n      Projection: #j1.key, #j1.c1\
            \n        TableScan: j1 projection=None\
            \n    Projection: #j2.key, #j2.c2\
            \n      TableScan: j2 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);

        let plan = LogicalPlanBuilder::from(
            LogicalPlanBuilder::from(make_sample_table("j1", vec!["key", "c1"])?)
                .project(vec![col("key"), col("c1")])?
                .build()?,
        )
        .join(
            &LogicalPlanBuilder::from(make_sample_table("j2", vec!["key", "c2"])?)
                .project(vec![col("key"), col("c2")])?
                .build()?,
            JoinType::Inner,
            (
                vec![Column::from_name("key")],
                vec![Column::from_name("key")],
            ),
        )?
        .project(vec![col("j1.c1"), col("j2.c2")])?
        .sort(vec![sort(col("j2.c2"), true, false)])?
        .build()?;

        let expected = "\
              Projection: #j1.c1, #j2.c2\
            \n  Sort: #j2.c2 ASC NULLS LAST\
            \n    Inner Join: #j1.key = #j2.key\
            \n      Projection: #j1.key, #j1.c1\
            \n        TableScan: j1 projection=None\
            \n      Projection: #j2.key, #j2.c2\
            \n        TableScan: j2 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);

        Ok(())
    }

    #[test]
    fn test_sort_down_cross_join() -> Result<()> {
        let plan = LogicalPlanBuilder::from(
            LogicalPlanBuilder::from(make_sample_table("j1", vec!["key", "c1"])?)
                .project(vec![col("key"), col("c1")])?
                .build()?,
        )
        .cross_join(
            &LogicalPlanBuilder::from(make_sample_table("j2", vec!["key", "c2"])?)
                .project(vec![col("key"), col("c2")])?
                .build()?,
        )?
        .project(vec![col("j1.c1"), col("j2.c2")])?
        .sort(vec![sort(col("j1.c1"), true, false)])?
        .build()?;

        let expected = "\
              Projection: #j1.c1, #j2.c2\
            \n  CrossJoin:\
            \n    Sort: #j1.c1 ASC NULLS LAST\
            \n      Projection: #j1.key, #j1.c1\
            \n        TableScan: j1 projection=None\
            \n    Projection: #j2.key, #j2.c2\
            \n      TableScan: j2 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);

        let plan = LogicalPlanBuilder::from(
            LogicalPlanBuilder::from(make_sample_table("j1", vec!["key", "c1"])?)
                .project(vec![col("key"), col("c1")])?
                .build()?,
        )
        .cross_join(
            &LogicalPlanBuilder::from(make_sample_table("j2", vec!["key", "c2"])?)
                .project(vec![col("key"), col("c2")])?
                .build()?,
        )?
        .project(vec![col("j1.c1"), col("j2.c2")])?
        .sort(vec![sort(col("j2.c2"), true, false)])?
        .build()?;

        let expected = "\
              Projection: #j1.c1, #j2.c2\
            \n  Sort: #j2.c2 ASC NULLS LAST\
            \n    CrossJoin:\
            \n      Projection: #j1.key, #j1.c1\
            \n        TableScan: j1 projection=None\
            \n      Projection: #j2.key, #j2.c2\
            \n        TableScan: j2 projection=None\
        ";

        assert_optimized_plan_eq(plan, expected);

        Ok(())
    }
}
