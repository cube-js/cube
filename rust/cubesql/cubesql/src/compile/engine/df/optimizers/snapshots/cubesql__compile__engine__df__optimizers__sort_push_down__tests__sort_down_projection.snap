---
source: cubesql/src/compile/engine/df/optimizers/sort_push_down.rs
expression: optimize(&plan)
---
Projection: #t1.c1 AS n1, #t1.c2, #t1.c3 AS n2, alias=t2
  Sort: #t1.c2 ASC NULLS LAST, #t1.c3 DESC NULLS FIRST
    Projection: #t1.c1, #t1.c2, #t1.c3
      TableScan: t1 projection=None
